src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RSA.cs

```diff
@@ -293,139 +288,75 @@ public virtual bool VerifyData(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signa
+                    return pkcs1PrivateKey.TryEncode(destination, out bytesWritten);
+                }
+            }
+            finally
+            {
+                ClearPrivateParameters(rsaParameters);
             }
         }
 
         public virtual byte[] ExportRSAPublicKey()
         {
-            using (AsnWriter pkcs1PublicKey = WritePkcs1PublicKey())
+            RSAParameters rsaParameters = ExportParameters(false);
+
+            using (AsnWriter pkcs1PublicKey = RSAKeyFormatHelper.WritePkcs1PublicKey(rsaParameters))
             {
                 return pkcs1PublicKey.Encode();
             }
         }
 
         public virtual bool TryExportRSAPublicKey(Span<byte> destination, out int bytesWritten)
         {
-            using (AsnWriter pkcs1PublicKey = WritePkcs1PublicKey())
+            RSAParameters rsaParameters = ExportParameters(false);
+
+            using (AsnWriter pkcs1PublicKey = RSAKeyFormatHelper.WritePkcs1PublicKey(rsaParameters))
             {
                 return pkcs1PublicKey.TryEncode(destination, out bytesWritten);
             }
         }
 
         public override unsafe bool TryExportSubjectPublicKeyInfo(Span<byte> destination, out int bytesWritten)
         {
-            // The PKCS1 RSAPublicKey format is just the modulus (KeySize bits) and Exponent (usually 3 bytes),
-            // with each field having up to 7 bytes of overhead and then up to 6 extra bytes of overhead for the
-            // SEQUENCE tag.
-            //
-            // So KeySize / 4 is ideally enough to start.
-            int rentSize = KeySize / 4;
-
-            while (true)
-            {
-                byte[] rented = ArrayPool<byte>.Shared.Rent(rentSize);
-                rentSize = rented.Length;
-                int pkcs1Size = 0;
-
-                fixed (byte* rentPtr = rented)
-                {
-                    try
-                    {
-                        if (!TryExportRSAPublicKey(rented, out pkcs1Size))
-                        {
-                            rentSize = checked(rentSize * 2);
-                            continue;
-                        }
-
-                        using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
-                        {
-                            writer.PushSequence();
-                            WriteAlgorithmIdentifier(writer);
-                            writer.WriteBitString(rented.AsSpan(0, pkcs1Size));
-                            writer.PopSequence();
-
-                            return writer.TryEncode(destination, out bytesWritten);
-                        }
-                    }
-                    finally
-                    {
-                        CryptographicOperations.ZeroMemory(rented.AsSpan(0, pkcs1Size));
-                        ArrayPool<byte>.Shared.Return(rented);
-                    }
-                }
-            }
-        }
+            RSAParameters rsaParameters = ExportParameters(false);
```

Comment:
The inheritance design that currently exists in master says that an RSA subclass need only overwrite TryExportRSAPublicKey and TryExportRSAPrivateKey to also get SubjectPublicKeyInfo, PrivateKeyInfo, and EncryptedPrivateKeyInfo.  They'd be forced to provide a definition for ExportParameters since it's abstract, but they could write it in terms of one of the other methods (whichever was best suited to their platform).

While it was true that simply defining ExportParameters would make things work, this change makes it so that ExportParameters is the primitive, and changing that requires a caller to also override the SubjectPublicKeyInfo and PrivateKeyInfo versions.

Nothing currently takes advantage of that inheritance model, but it was definitely something which was intentional.

Reply:
I noticed this, but decidedly didn't take advantage of it to keep the PR smaller and slightly easier to review (and it ended up huge anyway). I'd prefer to revisit this part in a separate PR.

Reply:
I'd really rather not lose the feature as an interim change.  I think that returning to that model will put this file back in pretty much it's current (in master) state, so it's probably "leave the rest of the PR alone, and start undoing deletes from this file".

Reply:
Uh, I misunderstood you the first time and I didn't realize it's broken. There's a room for improvement for *SecureTransforms to avoid some of the decoding and re-encoding in couple of cases. I didn't realize I broke the TryExportRSAPublicKey/TryExportRSAPrivateKey use case. I will try to fix that.

> While it was true that simply defining ExportParameters would make things work, this change makes it so that ExportParameters is the primitive, and changing that requires a caller to also override the SubjectPublicKeyInfo and PrivateKeyInfo versions.

It doesn't require any additional override as far as I can see. You always have to implement ExportParameters because it's abstract, so the only down side is that if you provide optimized implementation of TryExportRSAPublicKey/TryExportRSAPrivateKey it is not used instead.

Reply:
Okay, fair point.  If Pkcs8 export calls ExportParameters(true) which calls ExportRSAPrivateKey then the cost was extra array allocations and more copies of the private key floating in memory.

I have to think on whether that seems worth it. Certainly, when the feature was written (a few weeks ago) I thought that there was more value in not doing the extra arrays and extra copies.  And if X509 didn't have shared source for Apple's asymmetric crypto classes we'd just be leaving everything alone.

Reply:
My way of thinking was ... ok, so currently:
- `TryExportPkcs8PrivateKey` calls `WritePkcs8PrivateKey`
- `WritePkcs8PrivateKey` calls `ArrayPool<byte>.Shared.Rent` (1st allocation)
- Then it calls `TryExportRSAPrivateKey`, which in turn calls `WritePkcs1PrivateKey` (2nd allocation as part of `AsnWriter` + one copy as part of `AsnWriter.TryEncode`)
- `WritePkcs1PrivateKey` in turn calls `ExportParameters` (3rd allocation)
- `WritePkcs8PrivateKey` then allocates `AsnWriter` and copies the Pkcs1 blob there (4th allocation + copy)

That's assuming that `ExportParameters` is the primitive and the optimal case in `WritePkcs8PrivateKey` where everything fits the buffer.

In my rewrite:
- `TryExportPkcs8PrivateKey` calls `ExportParameters` (1st allocation)
- It calls `RSAKeyFormatHelper.WritePkcs8PrivateKey` (2nd allocation as part of AsnWriter that is returned)
- `RSAKeyFormatHelper.WritePkcs8PrivateKey` calls `WritePkcs1PrivateKey` (3rd allocation)
- One copy as part of `writer.WriteOctetString(pkcs1PrivateKey.EncodeAsSpan());` when copying the Pkcs#1 data into the Pkcs#8 writer

So, in the current code it's one less allocation and copy. If you reimplement `ExportParameters` using `ExportRSAPrivateKey` it will definitely change the numbers, but I will have to think about it a bit more.

Reply:
Ok, that is valid point. My whole thinking revolved around the fact that all the current code paths called `ExportParameters` anyway, so it saved one copying and one loop (which in most cases would come down to one iteration)... Anyway, I will try to revert it to the previous model. It makes more sense to me now.

Reply:
I reverted most of the changes, so now it follows your original design and it is much fewer changes. One thing I noticed though is that both overloads of `ImportEncryptedPkcs8PrivateKey` still go through `ImportParameters` instead of `ImportRSAPrivateKey`. This is reasonably easy to fix, so I guess I will put it to my backlog.